### ISSUE TITLE
Light clients and downloading block bodies

### ISSUE THREAD

**Issue by tomaka** [polkadot-fellows/RFCs#19]
After having submitted a transaction, a light client generally wants to know when and where the transaction is included in the finalized chain. This is both for UI purposes and to know when to stop gossiping that transaction.

In order to know when a transaction is included, the light client needs to download the body of each block of the best chain and compare the transactions of the body with the one that it has submitted.

While this is not a problem right now, if block bodies become larger in the future, light clients might have to use a lot of bandwidth just to download block bodies. After all, block bodies is one of the scaling factors of the chain. The busier the chain, the bigger the block bodies.

While a light client could ask a full node to send back a proof that a transaction is in a certain block, the full node could simply lie and pretend that the transaction isn't there. There is no way to prove that a transaction is not in a certain block body without sending the entire block body.

I don't have a solution to that problem, which is why I'm just opening an issue.

**Comment by tomaka** [comment]
One way to solve the problem could be to switch `extrinsics_root` to using V1 of the trie: https://github.com/paritytech/substrate/blob/ece92d6b30ee7f9fbda7a84bee76f8ff6d2c92a1/frame/system/src/lib.rs#L150

If this is done, we could then add a networking request for light clients to ask full nodes for a Merkle proof of the extrinsics trie that only contains the hashes of the extrinsics.
This proof would be around `35 * k` to `40 * k` bytes, where `k` is the number of extrinsics. This seems like a reasonable size to me.

Switching the `extrinsics_root` to V1 would be a breaking change, but we could reuse the `state_version` field for the migration (and rename it to `system_version` for example) by setting it to `2`.

Another possibility could be to modify the extrinsics trie so that the trie keys are something like `concat(transaction_hash, index)`. This would make it possible for a full node to generate a very small Merkle proof that an extrinsic is or is not in the block body.

**Comment by seunlanlege** [comment]
> `concat(transaction_hash, index)`

Light clients don't know the transaction index, i think just using transaction hash as the key is fine.

**Comment by tomaka** [comment]
The light client doesn't need to know the transaction index.
When a light client wants to know whether a transaction of a certain hash is included, the full node would send to it a Merkle proof of all the nodes whose key starts with that hash.

**Comment by josepot** [comment]
I’d like to propose a simpler and IMHO more pragmatic alternative to what was suggested above.

Instead of modifying the networking protocol and changing the `extrinsics_root` trie version, we could introduce a new runtime storage map: `System.Extrinsics`.

This would be a per-block `StorageMap<Hash, u32>`, where:

* the **key** is the transaction hash (using the `Identity` hasher),
* the **value** is the index of the extrinsic within the block body.

The map would be updated on every block and would not have a fallback value. Allowing anyone to trustlessly determine whether a transaction was included in a block (and if so, at which position) without downloading the entire body.

This approach:

* avoids any breaking protocol or trie changes
* enables precise and efficient inclusion checks
* adds minimal overhead (I _think_?)
* and leverages existing infrastructure.

WDYT? 🙏

**Comment by burdges** [comment]
Is `extrinsics_root` there already?  It's some Merkle tree of the extrinsics in the block?  

At a high level, we could compute block hash using some pharsing/context/extrinsics aware MMR, which costs roughly the same as hashing the block, so then you can prove inclusion from blockhash.  Actually you could do this by simply switching block hashes to blake3, but then you need to prove a bit more.

Non-inclusion requires some structure, like maybe you require the extrinsics be ordered by type and hash within the block.

Alternatively, your PVF could fill a `static mut extrinsic_indexed_ordered_by_hash: Vec<Option<(Hash,Index)>>` which the PVF keeps only in memory, not in storage.  `on_initialize` allocates this `Vec` on `None` for exactly the number of extrinsics in the block.  `on_finalize` computes the root of a binary Merkle tree whose leaves are `extrinsic_indexes_ordered_by_hash`, panics if any element is still `None`, and then checks that this root appeared correctly in the block header.  Now you can prove non-inclusin by showing the two adjacent elements. 

In general, there is no reason to write ephemeral data to storage, just work with hashes and `static mut`s or thread local storage or whartever.

**Comment by josepot** [comment]
> Is `extrinsics_root` there already?

yes... but AFAIK it currently uses `sp_core::storage::StateVersion::V0`, so that doesn't help.

> It's some Merkle tree of the extrinsics in the block?

@tomaka suggested migrating it to using `sp_core::storage::StateVersion::V1` and also adding a new networking request to ask full nodes for a Merkle proof of the extrinsics. That works. However, that solution has some serious implications:

- it's a breaking change at the protocol level
- it requires a new network request
- it requires a new JSON-RPC [`chainHead` version](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead.html) just to make this functionality available to tools.
- It requires serious changes on both PolkadotSDK nodes and smoldot (at least: implement the new JSON-RPC APIs and implement the new network request protocol)

In other words: dApp developers would have to wait an eternity before we are able to check if a transaction is present in a block without having to download the whole body, as we are currently doing. Which is a huge waste of bandwidth, and it's a rather expensive operation.

The solution that I'm suggesting is a lot more affordable, and it doesn't require any breaking changes.

Also, libraries like PAPI could easily check in the metadata whether the storage entry that I'm suggesting is present. If it's not present, then we fallback to accessing the whole body (as we are currently doing), and if `System. Extrinsics` is present, then we use it.

> In general, there is no reason to write ephemeral data to storage, just work with hashes and `static mut`s or thread local storage or whartever.

We are already using many other ephemeral (and sometimes redundant) storage entries, such as:
- System.Number
- System.ParentHash
- System.Digest
- System.Events
- System.EventCount

Unless there is a real issue in terms of performance, which I really don't think it's going to be the case here, I don't see why it should be a problem to add a ephemeral storage entry like the `System.Extrinsics` that I'm suggesting.

**Comment by burdges** [comment]
There is likely a somewhat serious performance hit from using ephemeral storage entries, certianly they're being kepts in merkle trees that hash everything, but worse those trees get packed into the PoV data.

We've needed a major overhaul of storage since before Kusama launched though, because our ETH-like storage was always dogshit, so this is really not a new problem, and the road to changing it isn't going to be short.

NOMT should force some better storage abstractions.  If this means multiple storages, then one could do a non-hashing in-memory storage, which maybe makes porting the code easier.

"Actions" in JAM reqjuires multiple independent state roots, and hence multiple storages, so by then we should definitely be able to have add a non-hashing in-memory storage that looked like any other storage.  This assumes nobody does anything stupid like make the actors be seperate pieces of code with exactly one state root each, but that'd really be against the spirit of JAM.

**Comment by joepetrowski** [comment]
> This would be a per-block `StorageMap<Hash, u32>`, where:
> 
> - the key is the transaction hash (using the Identity hasher),
> - the value is the index of the extrinsic within the block body.

This isn't completely foolproof since two extrinsics in the same block can have the same hash. It seems the sections I wrote about this in the wiki has been deleted or I cannot find it (cc @filippoweb3), but imagine:

- Account A has a balance of 100 DOT and nonce of 0.
- You have a block with:
    - A does `tranfer_all` to B (account A is reaped).
    - Someone else transfers some balance to A.
    - A does `tranfer_all` to B (account A is reaped).

Transactions `0` and `2` will have the same hash, but different indices (this is why we recommend using `(block_hash, extrinsic_index)` as the unique identifier for transactions, and _not_ the transaction hash). Of course, If you are looking for a certain _instance_ of a transaction, the hash is a valid identifier.

The options of storing a `Vec` either of transaction hashes or a concatenated `Vec<u8>` solve this by just neglecting the index, but generally we don't like unbounded vecs in the runtime (and it will be nasty for the light client to iterate through anyway). What about a valueless `StorageMap` with transaction hashes as the keys and `None`/`null` values?

**Comment by josepot** [comment]
> Since the hashes have a fixed-size of 32 bytes, it wouldn't even be necessary to use a Vector of Hashes, the hashes could be concatenated... So, it could just be a `StorageValue` of a `Vec<u8>`... That would spare us the bytes for the compact value of size, b/c it's redundant.

Ok, I just realized that this is silly, b/c in the end we would end up encoding the number of bytes which is greater than the number of hashes... So, a Vector of hashes is def better.

**Comment by bkchr** [comment]
> Switching the `extrinsics_root` to V1 would be a breaking change, but we could reuse the `state_version` field for the migration (and rename it to `system_version` for example) by setting it to `2`.

This is prepared and we could enact the switch on the runtime side. I don't remember when we merged this and we would need to check the number of nodes running with an old version that would break because of this.
